test/PushNotificationTests/TestFunctions.cpp

```diff
@@ -0,0 +1,151 @@
+
+    result.Channel().Close();
+    VERIFY_THROWS_HR(result.Channel().Close(), WPN_E_CHANNEL_CLOSED);
+}
+
+void PushNotificationTestFunctions::VerifyRegisterAndUnregister()
+{
+    auto pushNotificationManager{ PushNotificationManager::Default() };
+    pushNotificationManager.Register();
+    pushNotificationManager.Unregister();
+}
+
+void PushNotificationTestFunctions::VerifyRegisterAndUnregisterAll()
+{
+    auto pushNotificationManager{ PushNotificationManager::Default() };
+    pushNotificationManager.Register();
+    pushNotificationManager.UnregisterAll();
+}
+
+void PushNotificationTestFunctions::MultipleRegister()
+{
+    auto pushNotificationManager{ PushNotificationManager::Default() };
+    pushNotificationManager.Register();
+    VERIFY_THROWS_HR(pushNotificationManager.Register(), HRESULT_FROM_WIN32(ERROR_ALREADY_REGISTERED));
+}
+
+void PushNotificationTestFunctions::VerifyMultipleRegisterAndUnregister()
+{
+    PushNotificationManager::Default().Register();
+    PushNotificationManager::Default().Unregister();
+    VERIFY_THROWS_HR(PushNotificationManager::Default().Register(), HRESULT_FROM_WIN32(ERROR_ALREADY_REGISTERED));
+}
+
+void PushNotificationTestFunctions::VerifyMultipleRegisterAndUnregisterAll()
+{
+    PushNotificationManager::Default().Register();
+    PushNotificationManager::Default().UnregisterAll();
+
+    PushNotificationManager::Default().Register();
+    PushNotificationManager::Default().UnregisterAll();
+}
+
+void PushNotificationTestFunctions::VerifyUnregisterTwice()
+{
+    PushNotificationManager::Default().Register();
+    PushNotificationManager::Default().Unregister();
+    VERIFY_THROWS_HR(PushNotificationManager::Default().Unregister(), E_UNEXPECTED);
+}
+
+void PushNotificationTestFunctions::VerifyUnregisterAll()
+{
+    PushNotificationManager::Default().Register();
+    PushNotificationManager::Default().UnregisterAll();
+}
+
+void PushNotificationTestFunctions::VerifyUnregisterAllTwice()
+{
+    PushNotificationManager::Default().Register();
+    PushNotificationManager::Default().UnregisterAll();
+    VERIFY_THROWS_HR(PushNotificationManager::Default().UnregisterAll(), E_UNEXPECTED);
+}
+
+void PushNotificationTestFunctions::VerifyUnregisterAndUnregisterAll()
+{
+    PushNotificationManager::Default().Register();
+    PushNotificationManager::Default().Unregister();
+    PushNotificationManager::Default().UnregisterAll();
+}
+
+void PushNotificationTestFunctions::VerifyForegroundHandlerSucceeds()
+{
+    PushNotificationManager::Default().PushReceived([](const auto&, PushNotificationReceivedEventArgs const& /* args */) {});
+    PushNotificationManager::Default().Register();
+}
+
+void PushNotificationTestFunctions::VerifyForegroundHandlerFails()
+{
+    PushNotificationManager::Default().Register();
+    VERIFY_THROWS_HR(PushNotificationManager::Default().PushReceived([](const auto&, PushNotificationReceivedEventArgs const& /* args */) {}), HRESULT_FROM_WIN32(ERROR_NOT_FOUND));
```

Comment:
What wasn't found in this scenario so that call returns ERROR_NOT_FOUND?

Reply:
ERROR_NOT_FOUND is being thrown because there is no registration found. The THROW_HR_MSG will explain this error to the developer.

Reply:
Channel?

Reply:
> ERROR_NOT_FOUND is being thrown because there is no registration found.

App registration? But isn't calling Register() have taken care of it? This is pretty much the reason I initially started the comment thread. A comment in the code would be helpful.

Reply:
Checked it out again. ERROR_NOT_FOUND is being thrown because there is already a registration when trying to register an event handler, which goes against the pattern we have told developers. I don't think ERROR_NOT_FOUND is the correct error code to return to developers.

Reply:
Thank you @pmpurifoy for diving into it! Please file an issue to address this.

Reply:
https://github.com/microsoft/WindowsAppSDK/issues/2395